### ISSUE TITLE
Bug: Setting `env` to 'es2023' does not automatically enable es2023 syntax if extending a config that specifies its own `parserOptions.ecmaVersion`

### ISSUE THREAD

**Issue by polarstoat** [eslint/eslint#17679]
### Environment

Node version: v18.18.2
npm version: v9.8.1
Local ESLint version: v8.51.0 (Currently used)
Global ESLint version: v8.42.0
Operating System: darwin 22.5.0

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```js
module.exports = {
  env: {
    es2023: true,
  },
  extends: 'airbnb-base',
  // parserOptions: {
  //   ecmaVersion: '2023',
  // },
};
```
</details>

```js
// Somewhere in top-level of a module
function doSomethingAsync() {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve('foo');
    }, 300);
  });
}

const results = await doSomethingAsync();

console.log(results);
```

### What did you expect to happen?

The `env` config key being set with `es2023: true` to in turn automatically set `parserOptions.ecmaVersion` to `2023`, and therefore not encounter a parsing error with a top level await (introduced in es2022).

### What actually happened?

```  
/home/projects/stackblitz-starters-yqsarc/index.js
  10:17  error  Parsing error: Cannot use keyword 'await' outside an async function

✖ 1 problem (1 error, 0 warnings)
```

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/stackblitz-starters-yqsarc?file=.eslintrc.cjs

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

You can see by uncommenting the `parserOptions` key in my config that the parse error vanishes. The `eslint-config-airbnb-base` config internally uses `parserOptions: { ecmaVersion: 6 }` (see [line 6 of `es6.js`](https://github.com/airbnb/javascript/blob/11ab37144b7f846f04f64a29b5beb6e00d74e84b/packages/eslint-config-airbnb-base/rules/es6.js#L6) ) which appears to take precedence, despite the `env` being set in my derived config.

**Comment by mdjermanovic** [comment]
Hi @polarstoat, thanks for the issue!

This works as intended - `parserOptions` and `env` for a file are calculated independently from each other, and in the resulting config `parserOptions.ecmaVersion` always has precedence over `parserOptions.ecmaVersion` from environments, no matter where it was defined.

I agree this is confusing and often undesirable behavior. However, since the eslintrc config system is frozen, we are not making any changes to it at this point. In the new [flat config system](https://eslint.org/docs/latest/use/configure/configuration-files-new), there are no environments so this problem will not exist.
